src/ucp/proto/proto_perf.c

```diff
@@ -414,6 +414,67 @@ ucs_status_t ucp_proto_perf_aggregate(const char *name,
     return status;
 }
 
+ucs_status_t ucp_proto_perf_aggregate2(const char *name,
+                                       const ucp_proto_perf_t *perf1,
+                                       const ucp_proto_perf_t *perf2,
+                                       ucp_proto_perf_t **perf_p)
+{
+    const ucp_proto_perf_t *perf_elems[2] = {perf1, perf2};
+
+    return ucp_proto_perf_aggregate(name, perf_elems, 2, perf_p);
+}
+
+ucs_status_t ucp_proto_perf_remote(const ucp_proto_perf_t *remote_perf,
+                                   ucp_proto_perf_t **perf_p)
+{
+    ucp_proto_perf_segment_t *remote_seg, *new_seg;
+    ucp_proto_perf_factors_t perf_factors;
+    ucp_proto_perf_t *perf;
+    ucs_status_t status;
+
+    ucp_proto_perf_check(remote_perf);
+
+    status = ucp_proto_perf_create("remote", &perf);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    /* Turn local factors to remote and vice versa */
+    ucp_proto_perf_segment_foreach(remote_seg, remote_perf) {
+        perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_CPU] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_CPU];
+        perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_TL] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_TL];
+        perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_MTYPE_COPY] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_MTYPE_COPY];
+        perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY];
+        perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_MTYPE_COPY] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_MTYPE_COPY];
+        perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_TL] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_TL];
+        perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_CPU] =
+                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_CPU];
```

Comment:
How about using a static array of pairs?

Reply:
Not sure that I understand you correctly. Do you mean create two enums one is for factors and another for side?

Like:
```
typedef enum {
  UCP_PROTO_PERF_SIDE_LOCAL,
  UCP_PROTO_PERF_SIDE_REMOTE,
  UCP_PROTO_PERF_SIDE_LAST
} ucp_proto_perf_side_t;

typedef enum {
    UCP_PROTO_PERF_FACTOR_CPU,
    UCP_PROTO_PERF_FACTOR_TL,
    UCP_PROTO_PERF_FACTOR_MTYPE_COPY,
    UCP_PROTO_PERF_FACTOR_LATENCY,
    UCP_PROTO_PERF_FACTOR_LAST
} ucp_proto_perf_factor_id_t;
```

And then store them inside segment as two dimensional array?

```
typedef ucs_linear_func_t ucp_proto_perf_factors_t[UCP_PROTO_PERF_FACTOR_LAST][UCP_PROTO_PERF_SIDE_LAST];
```

Reply:
I mean two dimensional array: 3x2. And:
```C
for (i in size of array) {
    perf_factors[array[i][0]] = remote_seg->perf_factors[array[i][1]];
    perf_factors[array[i][1]] = remote_seg->perf_factors[array[i][0]];
}

perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY] = remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY];
```

Reply:
I found several ways how to do this operation instead of current way (option 0) and want to discuss with you what is the best option from your point of view. 

Let's start with your suggestion (option 1), if I don't miss something, it would look like the following:

```c
    static const size_t factors_to_convert = 3;
    ucp_proto_perf_factor_id_t convertaion_map[][2] = {
        {UCP_PROTO_PERF_FACTOR_LOCAL_CPU, UCP_PROTO_PERF_FACTOR_REMOTE_CPU},
        {UCP_PROTO_PERF_FACTOR_LOCAL_TL, UCP_PROTO_PERF_FACTOR_REMOTE_TL},
        {UCP_PROTO_PERF_FACTOR_LOCAL_MTYPE_COPY,
         UCP_PROTO_PERF_FACTOR_REMOTE_MTYPE_COPY}
    };

    ucp_proto_perf_segment_foreach(remote_seg, remote_perf) {
        for (factor_idx = 0; factor_idx < factors_to_convert; ++factor_idx) {
            perf_factors[convertaion_map[factor_idx][0]] =
                    remote_seg->perf_factors[convertaion_map[factor_idx][1]];
            perf_factors[convertaion_map[factor_idx][1]] =
                    remote_seg->perf_factors[convertaion_map[factor_idx][0]];
        }
        perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY] =
                remote_seg->perf_factors[UCP_PROTO_PERF_FACTOR_LATENCY];
        ...
    }
```

FMPOV it looks better inside the loop but I don't really like separate `factors_to_convert` variable. Another option (option 2) with defining `convertation_map` is using `ucs_swap`:

```c
    static const size_t factors_to_convert = 3;
    ucp_proto_perf_factor_id_t convertaion_map[][2] = {
        {UCP_PROTO_PERF_FACTOR_LOCAL_CPU, UCP_PROTO_PERF_FACTOR_REMOTE_CPU},
        {UCP_PROTO_PERF_FACTOR_LOCAL_TL, UCP_PROTO_PERF_FACTOR_REMOTE_TL},
        {UCP_PROTO_PERF_FACTOR_LOCAL_MTYPE_COPY,
         UCP_PROTO_PERF_FACTOR_REMOTE_MTYPE_COPY}
    };

    ucp_proto_perf_segment_foreach(remote_seg, remote_perf) {
        /* copy segments */
        for (fid = 0; fid < UCP_PROTO_PERF_FACTOR_LAST; fid++) {
            perf_factors[fid] = remote_seg->perf_factors[fid];
        }
        
        /* convert segments */
        for (factor_idx = 0; factor_idx < factors_to_convert; ++factor_idx) {
            ucs_swap(&perf_factors[convertaion_map[factor_idx][0]],
                     &perf_factors[convertaion_map[factor_idx][1]]);
        }
        ...
    }
```

Also I want to propose another solution (option 3) with `ucs_swap` but without `convertaion_map`:
```c
    ucp_proto_perf_segment_foreach(remote_seg, remote_perf) {
        /* copy segments */
        for (fid = 0; fid < UCP_PROTO_PERF_FACTOR_LAST; fid++) {
            perf_factors[fid] = remote_seg->perf_factors[fid];
        }
        
        /* convert segments */
        for (factor_idx = 0; factor_idx < factors_to_convert; ++factor_idx) {
            ucs_swap(&perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_CPU],
                     &perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_CPU]);
            ucs_swap(&perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_TL],
                     &perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_TL]);
            ucs_swap(&perf_factors[UCP_PROTO_PERF_FACTOR_LOCAL_MTYPE_COPY],
                     &perf_factors[UCP_PROTO_PERF_FACTOR_REMOTE_MTYPE_COPY]);
        }
        ...
    }
```

What do you think would be the best for this place?

P.S. I checked compilation of all these option, but didn't check correctness of the logic on practice. Despite this fact I think all of them should be valid.

Reply:
I think the first one is the simplest one. You can use `ucs_static_array_size` to avoid extra variable.

Reply:
Done